File: js/quick_sort.mjs
function quickSort(arr, left, right) {
    if(left <= right) {
        // 정렬되는 pivot의 Index를 반환
        let pivot = divide(arr, left, right);
        quickSort(arr, left, pivot - 1);
        quickSort(arr, pivot + 1, right);
    }
}

function divide(arr, left, right) {
    let pivot = arr[left];
    let leftStartIndex = left + 1;
    let rightStartIndex = right;

    while(leftStartIndex <= rightStartIndex) {
        while(leftStartIndex <= right && pivot >= arr[leftStartIndex]) {
            leftStartIndex++;
        }

        while(rightStartIndex >= (left + 1) && pivot <= arr[rightStartIndex]) {
            rightStartIndex--;
        }

        if(leftStartIndex <= rightStartIndex) {
            swap(arr, leftStartIndex, rightStartIndex);
        }
    }

    swap(arr, left, rightStartIndex)
    return rightStartIndex;
}

function swap(arr, index1, index2) {
    let temp = arr[index1];
    arr[index1] = arr[index2];
    arr[index2] = temp;
}

let arr = [3, 11, 6, 1, 2, 9, 20, 16, 31, 13]

console.log("=======정렬 전========")
console.log(arr);

quickSort(arr, 0, arr.length - 1);

console.log("=======정렬 후========")
console.log(arr);

File: python/QuickSort.py
def quick_sort(arr):
    if len(arr) <= 1:
        return arr
    
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

def quick_sort_in_place(arr):
    def sort(left, right):
        if left >= right:
            return
        
        mid = partition(left, right)
        sort(left, mid - 1)
        sort(mid, right)
        
    def partition(left, right):
        pivot = arr[(left + right) // 2]
        while left <= right:
            while arr[left] < pivot:
                left += 1
            while arr[right] > pivot:
                right -= 1
            if left <= right:
                arr[left], arr[right] = arr[right], arr[left]
                left += 1
                right -= 1
        return left
            
    return sort(0, len(arr) - 1)

sample_data = [38, 27, 43, 3, 9, 82, 10]
sorted_data1 = quick_sort(sample_data)
quick_sort_in_place(sample_data)
print(sorted_data1)
print(sample_data)